File: src/vertical_control/scripts/pg_ella/calcThetaStar.py
# -*- coding: utf-8 -*-
"""
Created on Tue Aug 11 12:45:38 2015

@author: brownyoda
"""

import numpy as np

from obtainData import obtainData
from episodicREINFORCE import episodicREINFORCE
from episodicNaturalActorCritic import episodicNaturalActorCritic
import sys

def calcThetaStar(Params, Policies, rates, trajlength, rollouts, numIterations):

	nSystems = np.shape(Params)[0]

	for i in range(nSystems):
		# TODO: Clear screen

		policy = Policies[i].policy # Resetting policy IMP
		for k in range(numIterations):
			print "@ Iteration: ", k
			data = obtainData(policy, trajlength, rollouts, Params[i])
			dJdtheta = None
			if Params[i].param.baseLearner == "REINFORCE":
				dJdtheta = episodicREINFORCE(policy, data, Params[i])
			else:
				dJdtheta = episodicNaturalActorCritic(policy, data, Params[i]) # TODO: won't use but should finish

			policy.theta = policy.theta + rates*dJdtheta.conj().T

		Policies[i].policy = policy # Calculating theta
		print "Policy ", i, ": ", policy.theta
	return Policies

File: CrossDomainPGELLAv1.0/calcThetaStar.m
function [Policies] = calcThetaStar(Params,Policies,rates,trajlength,rollouts,numIterations)

nSystems = size(Params,2); 

for i = 1:nSystems
    disp(['Calculating theta* for task ',num2str(i),': ',Params(i).param.ProblemID,'...',])

    policy = Policies(i).policy; % Resetting policy IMP  
    h = waitbar(0,['Please wait until theta* is calculated for task ',' ',num2str(i),': ',Params(i).param.ProblemID]);
    for k = 1:numIterations
        [data] = obtainData(policy, trajlength,rollouts,Params(i));
        dJdtheta = episodicNaturalActorCritic(policy,data,Params(i));
        policy.theta = policy.theta + rates*dJdtheta;
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % Uncomment this section to plot the reward function
        R = rand;
        G = rand;
        B = rand;
        for z = 1:size(data,2) % Loop over rollouts 
            r(z,:) = sum(data(z).r);
        end
        figure(i)
        plot(k,mean(r),'*','Color',[R,G,B])
        hold on 
        drawnow;            
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    
        waitbar(k / numIterations)
    end       
    close(h) 

    Policies(i).policy = policy; % Calculating theta*
end